test/cpp/api/tensor_indexing.cpp

```diff
@@ -0,0 +1,88 @@
+#include <gtest/gtest.h>
+
+// TODO: Move the include into `ATen/ATen.h`, once C++ tensor indexing
+// is ready to ship.
+#include <ATen/native/TensorIndexing.h>
+#include <torch/torch.h>
+
+#include <test/cpp/api/support.h>
+
+using namespace torch::indexing;
+using namespace torch::test;
+
+TEST(TensorIndexingTest, Slice) {
+  Slice slice(1, 2, 3);
+  ASSERT_EQ(slice.start(), 1);
+  ASSERT_EQ(slice.stop(), 2);
+  ASSERT_EQ(slice.step(), 3);
+
+  ASSERT_EQ(c10::str(slice), "{1, 2, 3}");
+}
+
+TEST(TensorIndexingTest, TensorIndex) {
+  {
+    std::vector<TensorIndex> indices = {None, "...", Ellipsis, 0, true, {1, None, 2}, torch::tensor({1, 2})};
+    ASSERT_TRUE(indices[0].is_none());
+    ASSERT_TRUE(indices[1].is_ellipsis());
+    ASSERT_TRUE(indices[2].is_ellipsis());
+    ASSERT_TRUE(indices[3].is_integer());
+    ASSERT_TRUE(indices[3].integer() == 0);
+    ASSERT_TRUE(indices[4].is_boolean());
+    ASSERT_TRUE(indices[4].boolean() == true);
+    ASSERT_TRUE(indices[5].is_slice());
+    ASSERT_TRUE(indices[5].slice().start() == 1);
+    ASSERT_TRUE(indices[5].slice().stop() == INDEX_MAX);
```

Comment:
Interesting. You wrote `None` in the slice specification, but it turns into `INDEX_MAX`?  I guess this is sound because you can never out of bound slice ranges in PYthon.